File: Spark-Essentials-TE2/Spark-Essentials/Labs/Optional/streaming/lab03/Scala/Lab.scala
// Databricks notebook source exported at Tue, 17 Nov 2015 18:05:53 UTC
// MAGIC %md
// MAGIC # Log File Streaming Lab
// MAGIC 
// MAGIC In this lab, we'll be reading (fake) log data from a stream.

// COMMAND ----------

import org.apache.spark._
import org.apache.spark.streaming._
import org.apache.spark.sql._
import scala.util.Random
import java.sql.Timestamp
import com.databricks.training.helpers.uniqueIDForUser

require(sc.version.replace(".", "").toInt >= 140, "Spark 1.4.0 or greater is required.")

// COMMAND ----------

val BatchIntervalSeconds= 1
val UniqueID = uniqueIDForUser()
val OutputDir = s"dbfs:/tmp/streaming/logs/$UniqueID"
val OutputFile = s"$OutputDir/${(new java.util.Date).getTime}.parquet"
val CheckpointDir = s"$OutputDir/checkpoint"


// COMMAND ----------

// MAGIC %md
// MAGIC We need to pull in some configuration data.
// MAGIC *In the next cell, EITHER change the path so it is an absolute path (starting with / up to the correct notebook)
// MAGIC OR just copy and paste the configuration cell*

// COMMAND ----------

val LogServerHost = "52.26.184.74"
val LogServerPort = 9001

// COMMAND ----------

// MAGIC %md 
// MAGIC ## Create our Streaming Process
// MAGIC 
// MAGIC We're going to connect to a TCP server that serves log messages, one per line. We'll read the messages and write them to a Parquet file.
// MAGIC 
// MAGIC We're going to be using a complicated-looking regular expression to take each log message apart:
// MAGIC 
// MAGIC ```
// MAGIC ^\[(.*)\]\s+\(([^)]+)\)\s+(.*)$
// MAGIC ```

// COMMAND ----------

// MAGIC %md
// MAGIC The following object contains the code for our stream reader.
// MAGIC 
// MAGIC **NOTE**: In Databricks (and in a Scala REPL), it helps to wrap your streaming solution in an object, to prevent scope confusion when Spark attempts to gather up and serialize the variables and functions for distribution across the cluster.

// COMMAND ----------

object Runner extends Serializable {
  import java.text.SimpleDateFormat
  import java.util.Date
  import java.sql.Timestamp
  
  case class LogMessage(messageType: String, timestamp: Timestamp, message: String)

  // Log messages look like this: [2015/09/10 21:15:46.680] (ERROR) Server log backup FAILED. See backup logs.
  private val LogLinePattern = """^\[(.*)\]\s+\(([^)]+)\)\s+(.*)$""".r
  
  /** Stop the streaming context.
    */
  def stop() {
    StreamingContext.getActive.map { ssc =>
      ssc.stop(stopSparkContext=false, stopGracefully=false)
      println("Stopped active streaming context.")
    }
  }
  
  /** Start the streaming context.
    */
  def start() {
    val ssc = StreamingContext.getActiveOrCreate(CheckpointDir, createStreamingContext)
    ssc.start()
    println("Started streaming context.")
  }
  
  /** Restart the streaming context (convenience method).
    */
  def restart() {
    stop()
    start()
  }
  
  // This method creates and initializes our stream.
  private def createStreamingContext(): StreamingContext = {
    val ssc = new StreamingContext(sc, Seconds(BatchIntervalSeconds))
    ssc.checkpoint(CheckpointDir)
    val DateFmt = new SimpleDateFormat("yyyy/MM/dd HH:mm:ss.S")

    val ds = ssc.socketTextStream(LogServerHost, LogServerPort)

    // What we're getting is a stream of log lines. As each RDD is created by
    // Spark Streaming, have the stream create a new RDD that extracts just
    // the error messages, dropping the rest.
    val keepLogsStream = ds.flatMap { line =>
      line match {
        case LogLinePattern(timeString, messageType, message) => {
          // Valid log line. Parse the date. NOTE: If this were a real
          // application, we would want to catch errors here.
          val timestamp = new Timestamp(DateFmt.parse(timeString).getTime)
          val logMessage = LogMessage(messageType, timestamp, message)
          Some(logMessage)
        }
        
        case _ => None // malformed line
      }
    }

    // Now, each RDD created by the stream will be converted to something
    // we can save to a persistent store that can easily be read, later,
    // into a DataFrame.
    keepLogsStream.foreachRDD { rdd =>
      if (! rdd.isEmpty) {
        // Using SQLContext.createDataFrame() is preferred here, since toDF()
        // can pick up (via implicits) objects that aren't serializable.
        val sqlContext = SQLContext.getOrCreate(SparkContext.getOrCreate())
        val df = sqlContext.createDataFrame(rdd)
        df.write.mode(SaveMode.Append).parquet(OutputFile)
      }
    }
        
    // Start our streaming context.
    ssc.start()

    ssc
  }
}

// COMMAND ----------

// MAGIC %md For good measure, let's make sure our `Runner` class is serializable. If it isn't, we'll get an error when we try to start it.

// COMMAND ----------

import java.io.{ByteArrayOutputStream, ObjectOutputStream}
val os = new ObjectOutputStream(new ByteArrayOutputStream)
os.writeObject(Runner)


// COMMAND ----------

dbutils.fs.rm(OutputDir, recurse=true)
dbutils.fs.mkdirs(OutputDir)
Runner.restart()


// COMMAND ----------

display( dbutils.fs.ls(OutputDir) )

// COMMAND ----------

// Is there anything in the output directory?
display( dbutils.fs.ls(OutputFile) )


// COMMAND ----------

// MAGIC %md
// MAGIC Let the stream run for awhile. Then, execute the following cell to stop the stream.

// COMMAND ----------

Runner.stop()

// COMMAND ----------

// MAGIC %md Okay, now let's take a look at what we have in the Parquet file. **NOTE**: This next command may fail periodically if the stream is still running.

// COMMAND ----------

val df = sqlContext.read.parquet(OutputFile)

// COMMAND ----------

// MAGIC %md How many messages did we pull down?

// COMMAND ----------

df.count

// COMMAND ----------

display(df)

// COMMAND ----------

df.registerTempTable("messages")

// COMMAND ----------

// MAGIC %md
// MAGIC ## Exercise 1
// MAGIC 
// MAGIC How many messages of each type (ERROR, INFO, WARN) are there?
// MAGIC 
// MAGIC **HINT**: Use your DataFrame knowledge.

// COMMAND ----------

// MAGIC %sql select messageType, count(messageType) from messages group by messageType

// COMMAND ----------

// MAGIC %md 
// MAGIC ## Exercise 2
// MAGIC 
// MAGIC Modify the `Runner`, above, to keep only the ERROR messages.

// COMMAND ----------

display(dbutils.fs.ls("dbfs:/tmp/streaming"))

// COMMAND ----------

File: Instructors Area/Instructor-Materials/Spark-Essentials/Labs/Optional/streaming/lab03/Python/Lab.py
# Databricks notebook source exported at Sat, 19 Sep 2015 14:06:06 UTC
# MAGIC %md
# MAGIC # Log File Streaming Lab
# MAGIC 
# MAGIC In this lab, we'll be reading (fake) log data from a stream.

# COMMAND ----------

# MAGIC %md
# MAGIC 
# MAGIC ## WARNING
# MAGIC 
# MAGIC This lab does not currently work. There are, apparently, issues with the Python streaming library.

# COMMAND ----------

from pyspark import *
from pyspark.streaming import *
from pyspark.sql import *
import random
from datetime import datetime

def now_as_millis():
  import time
  import calendar
  return calendar.timegm(time.gmtime())

assert int(sc.version.replace(".", "")) >= 140, "Spark 1.4.0 or greater is required."

# COMMAND ----------

BATCH_INTERVAL_SECONDS = 15
OUTPUT_DIR = "dbfs:/tmp/streaming/logs/{0}".format(random.randint(1, 1000))
OUTPUT_FILE = "{0}/{1}.parquet".format(OUTPUT_DIR, now_ms)
CHECKPOINT_DIR = "{0}/checkpoint".format(OUTPUT_DIR)

print "Using: Batch interval of {0} second(s)".format(BATCH_INTERVAL_SECONDS)
print '       Output directory: "{0}"'.format(OUTPUT_DIR)
print '       Output file:      "{0}"'.format(OUTPUT_FILE)
print '       Checkpoint dir:   "{0}"'.format(CHECKPOINT_DIR)


# COMMAND ----------

# MAGIC %md
# MAGIC We need to pull in some configuration data.

# COMMAND ----------

# MAGIC %run "/Labs/streaming/lab03/Python/Configuration"

# COMMAND ----------

# MAGIC %md 
# MAGIC ## Create our Streaming Process
# MAGIC 
# MAGIC We're going to connect to a TCP server that serves log messages, one per line. We'll read the messages and write them to a Parquet file.
# MAGIC 
# MAGIC We're going to be using a complicated-looking regular expression to take each log message apart:
# MAGIC 
# MAGIC ```
# MAGIC ^\[(.*)\]\s+\(([^)]+)\)\s+(.*)$
# MAGIC ```

# COMMAND ----------

# MAGIC %md
# MAGIC The following class contains the code for our stream reader.

# COMMAND ----------

import re
from collections import namedtuple

class Runner(object):

  def __init__(self):
    self._ssc = None
    
  def stop(self):
    """
    Stop the streaming context.
    """
    if self._ssc:
      self._ssc.stop(stopSparkContext=False, stopGraceFully=True)
      self._ssc = None
      print "Stopped active streaming context."
  
  def start(self):
    """
    Start the streaming context
    """
    # NOTE: StreamingContext.getOrCreate() exists in the Python API, but it currently
    # throws errors.
    #self._ssc = StreamingContext.getOrCreate(CHECKPOINT_DIR, Runner._create_streaming_context)
    self._ssc = Runner._create_streaming_context()
    print "Starting {0}".format(self._ssc)
    self._ssc.start()
    print "Started streaming context."
  
  def restart(self):
    """
    Restart the streaming context.
    """
    self.stop()
    self.start()
   
  @classmethod
  def _create_streaming_context(cls):
    LogMessage = namedtuple('LogMessage', ['message_type', 'timestamp', 'message'])

    LOG_LINE_PATTERN = re.compile(r'^\[(.*)\]\s+\(([^)]+)\)\s+(.*)$')
    TIMESTAMP_FORMAT = '%Y/%m/%d %H:%M:%S.%f'
    
    ssc = StreamingContext(sc, BATCH_INTERVAL_SECONDS)
    ds = ssc.socketTextStream(LOG_SERVER_HOST, LOG_SERVER_PORT)

    # What we're getting is a stream of log lines. As each RDD is created by
    # Spark Streaming, have the stream create a new RDD that extracts just
    # the error messages, dropping the rest.
    def parse_log_line(line):
      log_message = None
      try:
        m = LOG_LINE_PATTERN.search(line)
        if m:
          timestamp = datetime.strptime(TIMESTAMP_FORMAT, m.group(1))
          log_message = LogMessage(m.group(2), timestamp, m.group(3))
      except:
        pass

      return log_message

    #keep_logs_stream = ds.map(parse_log_line).filter(lambda line: line is not None)

    # Now, each RDD created by the stream will be converted to something
    # we can save to a persistent store that can easily be read, later,
    # into a DataFrame.
    def save_rdd(rdd):
      print "*** processing rdd. rdd.isEmpty={0}".format(rdd.isEmpty())
      #if not rdd.isEmpty():
        #df = sqlContext.createDataFrame(rdd)
        #df.write.mode("append").parquet(OUTPUT_FILE)
      rdd.saveAsTextFile("{0}-{1}".format(OUTPUT_DIR, now_as_millis()))

    #keep_logs_stream.foreachRDD(save_rdd)
    ds.foreachRDD(save_rdd)

    print "--- ssc={0}".format(ssc)
    return ssc  

# COMMAND ----------

dbutils.fs.rm(OUTPUT_DIR, recurse=True)
dbutils.fs.mkdirs(OUTPUT_DIR)
runner = Runner()
runner.stop()
runner.restart()


# COMMAND ----------

# Is there anything in the output directory?
display( dbutils.fs.ls(OUTPUT_DIR) )


# COMMAND ----------

# MAGIC %md
# MAGIC Let the stream run for awhile. Then, execute the following cell to stop the stream.

# COMMAND ----------

runner.stop()

# COMMAND ----------

# MAGIC %md Okay, now let's take a look at what we have in the Parquet file. **NOTE**: This next command may fail periodically if the stream is still running.

# COMMAND ----------

val df = sqlContext.read.parquet(OUTPUT_FILE)

# COMMAND ----------

# MAGIC %md How many messages did we pull down?

# COMMAND ----------

df.count

# COMMAND ----------

display(df)

# COMMAND ----------

df.registerTempTable("messages")

# COMMAND ----------

# MAGIC %md
# MAGIC ## Exercise 1
# MAGIC 
# MAGIC How many messages of each type (ERROR, INFO, WARN) are there?
# MAGIC 
# MAGIC **HINT**: Use your DataFrame knowledge.

# COMMAND ----------

# MAGIC %sql select messageType, count(messageType) from messages group by messageType

# COMMAND ----------

# MAGIC %md 
# MAGIC ## Exercise 2
# MAGIC 
# MAGIC Modify the `Runner`, above, to keep only the ERROR messages.

# COMMAND ----------



# COMMAND ----------

